File: C++/Searches/binary_search.cpp
#include<iostream>
using namespace std;
#define maxSize 100;

int main()

    int arr[maxSize] , element , mid , start , end ;
    int arrSize , i , j , temp ;
    mid = start = end = 0;
    
    cout << " \n Enter size of array : " ;
    cin >> arrSize ;
    
    cout << " \n Enter elements of array: " ;
    for ( i = 0 ; i < n ; i++ ){

        cin >> arr[i] ;
    }
    
    // SORTING ARRAY 
    for ( i = 0 ; i < arrSize ; i++ ) {
 
        for ( j = 0 ; j < arrSize-i ; j++) {

            if ( arr[ j ] > arr[ j+1 ] ) {
                
                temp = arr[ j ] ;
                arr[ j ] = arr[ j+1 ];
                arr[ j+1 ] = temp;
            }
        }
    }

    cout << " \n Sorted array : " << endl ;
    
    for ( i = 0 ; i < arrSize ; i++ ) {
  
        cout << arr[ i ] << " " << endl ;
    }

    end  = arrSize-1 ;

    cout << " \n Enter an element to be searched : " ;
    cin >> element ;

    mid = ( start + end ) / 2 ;

    while ( start < end  &&  arr[ mid ] ! = element ) {
        if ( element > arr[ mid ] ) {
            start = mid+1;
        }
        else {
            end = mid - 1 ;
        }
        mid = ( start + end ) / 2 ;
    }

    if ( arr[ mid ] == element ) {
        cout << " \n Element " << arr[ mid ] << " found at " << mid+1 << " position " << endl ;
    }
    else {
        cout << " \n Element not found " << endl ;
    }
 
     return 0;
}

File: Kotlin/Searches/BinarySearch.kt
fun binarySearch(arr: IntArray, lowerIndex: Int, upperIndex: Int, x: Int): Int {
    if (upperIndex >= lowerIndex) {
        val mid = lowerIndex + (upperIndex - lowerIndex) / 2

        // If the element is found
        if (arr[mid] == x)
            return mid

        /* If the element is smaller than mid, then it can only be present in left subarray
        else the element can only be present in right subarray */
        return if (arr[mid] > x) binarySearch(arr, lowerIndex, mid - 1, x)
        else binarySearch(arr, mid + 1, upperIndex, x)
    }
    // We reach here only when the element is not present in array
    return -1
}

fun main(args: Array<String>) {
    println("Enter the size of array:")
    val n = readLine()!!.toInt()
    println("Enter the elements of array:")
    var arr = IntArray(n) { readLine()!!.toInt() }
    println("Enter a number to search:")
    val x = readLine()!!.toInt()
    arr = arr.sortedArray() // Sorting the array in ascending order
    println("Array in ascending order: ")
    for (l in arr)
        print("$l ")
    println()
    val result = binarySearch(arr, 0, n - 1, x)
    if (result == -1)
        println("Element not present")
    else
        println("Element found at index $result")
}